design/multiple-storage-types-support.md

```diff
@@ -0,0 +1,272 @@
+
+Custom resources in Kubernetes use the following naming and versioning convention:
+
+* Group: A collection of several related types that can be enabled/disabled as a unit in the API (e.g., `rook.io`)
+* Version: The API version of the group (e.g., `v1alpha1`)
+* Kind: The specific type within the API group (e.g., `cluster`)
+
+Putting this together with an example, the `cluster` kind from the `rook.io` API group with a version of `v1alpha1` would be referred to in full as `cluster.rook.io/v1alpha1`.
+
+#### StorageScope
+
+The `storageScope.rook.io/v1alpha2` type defines the boundaries or "scope" of the resources that comprise the backing storage substrate for a cluster.
+This could be devices, filters, directories, nodes, [persistent volumes](https://github.com/rook/rook/issues/919), and others.
+There are user requested means of selecting storage that Rook doesn't currently support that could be included in this type, such as the ability to select a device by path instead of by name, e.g. `/dev/disk/by-id/`.
+Also, wildcards/patterns/globbing should be supported on multiple resource types, removing the need for the current `useAllNodes` and `useAllDevices` boolean fields.
+
+By encapsulating this concept as its own type, it can be reused within other custom resources of Rook.
+For instance, this would enable Rook to support storage of other types that could benefit from orchestration in cloud-native environments beyond distributed storage systems.
+`StorageScope` could also provide an abstraction from such details as device name changes across reboots.
+
+#### StorageCluster
+
+The `storageCluster.rook.io/v1alpha2` type provides the general abstraction for a distributed storage cluster.
+The spec for this type will include image version, networking information, placement information, resource usage requests, etc.
+Backend specific fields, such as Ceph configuration, will be included under a field identified by the backend's name.
+Examples of the `StorageCluster` type are included in the [examples](#examples) section.
+
+In addition to the existing support for Ceph, we should consider support for [Minio](https://www.minio.io/) as the next storage backend supported by Rook.
+Minio is a popular distributed object storage solution that could benefit from a custom controller to orchestrate, configure and manage it in cloud-native environments.
+
+#### Additional Types
+
+Rook and the greater community would also benefit from additional types and abstractions.
+We should work on defining those further, but it is out of scope for this design document that is focusing on support for multiple storage backends.
+Some potential ideas for additional types to support in Rook:
+
+* [Snapshots, backup and policy](https://github.com/rook/rook/issues/1552)
+* Quality of Service (QoS), resource consumption (I/O and storage limits)
+
+## Examples
+
+This section contains concrete examples of storage clusters as a user would define them using yaml.
```

Comment:
I think we should weigh the pros and cons of the combined crd definition vs one per backend. For example, pros of individual crds include a clear versioning story, backend-specific settings can be specified inline for nodes/devices, and don't need a separate backend-specific section.

Reply:
The approach of one crd per backend would imply that Rook would benefit the backends at the source code level. The deserializing of types and framework around handling of CRDs, etc.

Reply:
Generally speaking "nodes" objects needs to know which "zones" they belong to and "devices" doesn't need to know .... because "nodes" would need to be selected based on matching criteria anyways. And "devices" are considered to be a resource which has to be allocated by K8s one way or the other. I've looked at upcoming Local Persistent Volumes effort. I think it would be fine for now to be more explicit (perhaps requiring /dev/disk/by-id rather then /dev/sda because it is not persistent across reboots) in StorageScope definitions but we should not forget that soon we will be able to manage devices (filesystems, partitions and raw) in more generic "resource-like" terms.

Reply:
Does anyone have an idea what the user experience would be like if [Local Storage](https://kubernetes.io/docs/concepts/storage/volumes/#local) were to be used?  It's now in Beta for k8s 1.10, but I'm struggling understanding how it would be integrated smoothly into Rook.  What would using LocalStorage PV's within Rook's `StorageScope` type look like for a user?  Does it integrate with Rook in a different way?

Reply:
Storage Clusters (including NexentaEdge) would prefer to manage persistence themselves based on having been given resources. The storage resource should be anything suitable to place a file system on (i.e., a virtual disk file, a partition or a jbod.